### PR TITLE
feat: Add conversion webhook for ArgoCD v1alpha1 to v1beta1 migration

### PR DESCRIPTION
**Pull request by svghadi**
**What type of PR is this?**
/kind enhancement

**What does this PR do / why we need it**:
This PR deprecates ArgoCD `v1alpha` CRD and adds a new ArgoCD `v1beta1` CRD along with conversion webhook to automate migration of ArgoCD v1alpha1 resources to v1beta1.

Changes:
- Add new ArgoCD v1beta1 crd.
- Mark ArgoCD v1alpha1 as deprecated & add back the removed dex & sso deprecated fields in ArgoCD v1alpha1.
- Add explicit conversion funcs for converting  ArgoCD v1alpha1 to v1beta1.
- Configure conversion webhook server to run on 9443 port in operator-controller-manager deployment.
- Disable single namespace install for operator. OLM doesn't support single namespace installation mode for operators with conversion webhooks.
- Use server side validation for "kubectl apply" as client side apply results into failure due to exceeding annotation size limit.
- Add kuttl e2e tests
- Add documentation for manual install of operator with conversion webhook support.

Follow-up PRs (yet to be created):
- Add CI job for testing conversion webhook
- Replace all ArgoCD v1alpha1 references with v1beta1 in codebase
- Update api docs

**Have you updated the necessary documentation?**

* [x] Documentation update is required by this PR.
* [x] Documentation has been updated. (**Partially**)

**How to test changes / Special notes to the reviewer**:
Operator Installation:
- Follow `tests/olm/README.md` from this PR to install the operator via OLM. 

Manual Test:
- Create a ArgoCD with v1alpha1 version 
    ```bash
    kubectl apply -f - <<EOF
    apiVersion: argoproj.io/v1alpha1
    kind: ArgoCD
    metadata:
      name: argocd
    spec:
      dex:
        config: test-config
    EOF
    ```
- Observe deprecation warning for `kubectl apply`
    `Warning: ArgoCD v1alpha1 is deprecated, please use v1beta1 instead.`

- Verify that conversion to v1beta1 didn't result in data loss
    ```bash
    $ kubectl get argocd/argocd -o yaml
    apiVersion: argoproj.io/v1beta1
    kind: ArgoCD
    metadata:
      annotations:
        kubectl.kubernetes.io/last-applied-configuration: |
          {"apiVersion":"argoproj.io/v1alpha1","kind":"ArgoCD","metadata":{"annotations":{},"name":"argocd","namespace":"default"},"spec":{"dex":{"config":"test-config"}}}
      name: argocd
    spec:
      sso:
        dex:
          config: test-config
        provider: dex
    ```

Automated Test:
- `kubectl kuttl test ./tests/olm --config ./tests/kuttl-tests.yaml`

### CODE REVIEW

File: api/v1alpha1/argocd_conversion.go
Comment (jaideepr97) [inline api/v1alpha1/argocd_conversion.go:82]: aren't we dropping `Spec.ResourceCustomizations` in the `v1beta1` spec as well?

File: api/v1beta1/argocd_types.go
Comment (jaideepr97) [inline api/v1beta1/argocd_types.go:753]: I think we want to get rid of this, as mentioned in the earlier comment

File: config/default/webhookcainjection_patch.yaml
Comment (jaideepr97) [inline config/default/webhookcainjection_patch.yaml:15]: are we introducing a validating webhook as a part of this PR as well?

File: deploy/olm-catalog/argocd-operator/0.7.0/argoproj.io_argocds.yaml
Comment (jaideepr97) [inline deploy/olm-catalog/argocd-operator/0.7.0/argoproj.io_argocds.yaml:1]: I think we probably want to keep these changes out of the 0.7.0 bundle 
they should be in the 0.8.0 bundle when we have one

**Comment by jaideepr97** [comment]
awesome work @svghadi :tada: 

Thinking a little future facing, we currently have  `convertFrom` and `convertTo` functions to move between alpha1/beta1
Is that enough to support future use cases when we may have a beta2/beta3 as well, without many changes?

**Comment by jaideepr97** [comment]
Also, I think we will need to go into the operator and update all the `v1alpha1` references to `v1beta1` in the reconciler logic, either in this PR or a follow up one

File: Makefile
Comment (iam-veeramalla) [inline Makefile:113]: Can you confirm the same sed command works on both mac and linux ? sometimes there is a little syntaxing issues.

File: Makefile
Comment (iam-veeramalla) [inline Makefile:216]: I hope you verified the compatibility with this new version.

File: api/v1alpha1/argocd_conversion.go
Comment (iam-veeramalla) [inline api/v1alpha1/argocd_conversion.go:27]: deprecated fields will have more priority ?

File: api/v1alpha1/argocd_conversion.go
Comment (svghadi) [reply api/v1alpha1/argocd_conversion.go:82]: Yes. I will get those changes in a separate PR.

File: api/v1beta1/argocd_types.go
Comment (svghadi) [reply api/v1beta1/argocd_types.go:753]: Will remove it in a separate PR.

File: config/default/webhookcainjection_patch.yaml
Comment (svghadi) [reply config/default/webhookcainjection_patch.yaml:15]: No, only conversion webhook is added. 

This configuration is created automatically via `operator-sdk create` command during webhook creation. It is created to allow cert-manager to inject certs for validating & mutating webhook.  Since we don't have them in operator and by default it is disabled, it can be ignored as it is not added in final `kustomize build`.

File: deploy/olm-catalog/argocd-operator/0.7.0/argoproj.io_argocds.yaml
Comment (svghadi) [reply deploy/olm-catalog/argocd-operator/0.7.0/argoproj.io_argocds.yaml:1]: Agree.

 @reginapizza - Now that argocd-operator `0.7.0` is out, should we set up master for `0.8.0` or is it too early?

**Comment by svghadi** [comment]
> Thinking a little future facing, we currently have convertFrom and convertTo functions to move between alpha1/beta1
Is that enough to support future use cases when we may have a beta2/beta3 as well, without many changes?  

Yes. Whenever we add a new version, we have to decide on which version to store in etcd(hub version) and define `convertFrom` & `convertTo` for other version(spokes) along with few configuration changes.

> Also, I think we will need to go into the operator and update all the v1alpha1 references to v1beta1 in the reconciler logic, either in this PR or a follow up one

Yes. Since there will be lot of files changed, I didn't make it in this PR as it will be difficult to review. I will create follow up PR for this.

> And probably think about what we need to do to support the webhook installation via non-OLM installation as well

For non-OLM install, the webhook will work as expected **provided** user handles certificate generation and injection which can be automated on k8s using `cert-manager` and on openshift using `Service CA Operator`. I have updated the installation docs with steps for the same in this PR.

File: Makefile
Comment (svghadi) [reply Makefile:113]: Yes. It works on both mac & linux.

File: Makefile
Comment (svghadi) [reply Makefile:216]: I didn't find any compatibility matrix for opm and operator-sdk or olm. However I did use this new version to create registry image for operator testing, didn't see any issues while creation or running the registry image. 

Also this change has nothing to do with the conversion webhook or crd upgrade. I made it for my local test environment as older version didn't support mac(arm). If needed, we can always revert this change.

File: api/v1alpha1/argocd_conversion.go
Comment (svghadi) [reply api/v1alpha1/argocd_conversion.go:27]: Here, the idea of conversion webhook is for smooth migration of existing v1alpha1 users to v1beta1. If any existing users are using deprecated fields that means they have not updated their sso configurations to use new keys hence having more priority to deprecated fields in such cases makes more sense to me.

**Comment by iam-veeramalla** [comment]
> I am yet to complete the review on this PR, reviewed some of it. Will continue the review tomorrow.

Please don't block the PR on my review. I see @jaideepr97 is actively working on this one.

I am not available for the whole next week. @svghadi , please feel free to merge the PR if you have approval from Jaideep.

**Comment by svghadi** [comment]
> Instead of defining a new pair of convertFrom/convertTo functions each time we have a new version, would it make sense to have a single convertFrom/convertTo function pair, which takes in a non-hub version as input and then uses an interface (probably a map) to figure out which conversion functions should be applied?
That way we only have to define conversion functions for individual fields going forwards

@jaideepr97 - Can you please elaborate on this? Are you suggesting that we create a single, generic `convertFrom/convertTo` function? 

My understanding about spokes is that each spoke needs to implement the `Convertible` interface from kubebuilder for conversion, which means each spoke would have its own `convertFrom/convertTo` function.

**Comment by jaideepr97** [comment]
> My understanding about spokes is that each spoke needs to implement the Convertible interface from kubebuilder for conversion, which means each spoke would have its own convertFrom/convertTo function.

Ah okay
I also just realized that each `convertFrom/convertTo` will also be in its own package 
nevermind, this is fine

File: Dockerfile
Comment (svghadi) [reply Dockerfile:21]: With `GOARCH` hardcoded to `amd64`, the build command creates binary for amd64 on my mac(arm64) local machine.  If we don't specify anything, it will default to the host arch on which it builds.

Comment (svghadi) [reply Dockerfile:21]: We should update docker build commands in Makefile to use `--platform` flag to make cross-buidling of images easy.

File: api/v1alpha1/argocd_conversion.go
Comment (jaideepr97) [reply api/v1alpha1/argocd_conversion.go:27]: I think its fine either way, the current checks prevent a confict like that from happening in the first place 
i.e, they can't have the same fields with conflicting info in `.spec.dex` and `.spec.sso.dex` in their `v1alpha1` CRs anyway

File: docs/install/openshift.md
Comment (jaideepr97) [inline docs/install/openshift.md:46]: maybe we should mention that it is enabled out of the box
https://github.com/argoproj-labs/argocd-operator/pull/964/files#diff-29597131adc5aaecd3350041d098624e877013a910db6312107dadfaa0a5690eR1750

File: docs/install/openshift.md
Comment (jaideepr97) [inline docs/install/openshift.md:43]: By "manual" do we mean non-OLM here?

**Comment by jaideepr97** [comment]
@svghadi tested the PR and it works great :+1:  Please remove the changes from the 0.7.0 bundle for now
once the master branch is set up for 0.8.0, these changes will automatically get there, so I don't want to block this PR just for that

**Review by jaideepr97 — APPROVED** [review]

File: docs/install/openshift.md
Comment (svghadi) [reply docs/install/openshift.md]: Yes. I updated the docs to reflect this.

**Comment by svghadi** [comment]
@jaideepr97 - I have removed the changes from 0.7.0 bundle. However code gen CI job is failing due to this.

**Comment by jaideepr97** [comment]
@svghadi we should be having a PR to set up 0.8.0 soon, once that is merged please rebase your PR so we can get this merged as well

**Comment by svghadi** [comment]
Rebased the PR with 0.8.0 changes. It is ready for merge. Thanks

**Comment by abdennour** [comment]
i like it

**Pull request merged**
